File: CPP/Graphs/LCAForest.cpp
// Least Common Ancestor Forest
//
template <typename T>
class LCAForest : public DFSForest<T> {
  public:
	using DFSForest<T>::edges;
	using DFSForest<T>::g;
	using DFSForest<T>::n;
	using DFSForest<T>::pv;
	using DFSForest<T>::pos;
	using DFSForest<T>::end;
	using DFSForest<T>::depth;

	int h;
	vector<vector<int>> pr;

	LCAForest(int _n) : DFSForest<T>(_n) {
	}

	inline void build_lca() {
		assert(!pv.empty());
		int max_depth = 0;
		for (int i = 0; i < n; i++) {
			max_depth = max(max_depth, depth[i]);
		}
		h = 1;
		while ((1 << h) <= max_depth) {
			h++;
		}
		pr.resize(n);
		for (int i = 0; i < n; i++) {
			pr[i].resize(h);
			pr[i][0] = pv[i];
		}
		for (int j = 1; j < h; j++) {
			for (int i = 0; i < n; i++) {
				pr[i][j] = (pr[i][j - 1] == -1 ? -1 : pr[pr[i][j - 1]][j - 1]);
			}
		}
	}

	inline bool anc(int x, int y) {
		return (pos[x] <= pos[y] && end[y] <= end[x]);
	}

	inline int go_up(int x, int up) {
		assert(!pr.empty());
		up = min(up, (1 << h) - 1);
		for (int j = h - 1; j >= 0; j--) {
			if (up & (1 << j)) {
				x = pr[x][j];
				if (x == -1) {
					break;
				}
			}
		}
		return x;
	}

	inline int lca(int x, int y) {
		assert(!pr.empty());
		if (anc(x, y)) {
			return x;
		}
		if (anc(y, x)) {
			return y;
		}
		if (depth[x] > depth[y]) {
			swap(x, y);
		}
		for (int j = h - 1; j >= 0; j--) {
			if (pr[x][j] != -1 && !anc(pr[x][j], y)) {
				x = pr[x][j];
			}
		}
		return pr[x][0];
	}
};

File: Kotlin/Graphs/LcaForest.kt
open class LcaForest(n_:Int) : DfsForest(n_) {
    var h:Int = 0
    var pr = Array<IntArray>(0, {intArrayOf()})
    fun build_lca() {
        require(pv.isNotEmpty()){"Do DFS First"}
        var max_depth:Int = 0
        for(i in 0 until n) {
            max_depth = Math.max(max_depth, depth[i])
        }
        h = 1
        while ((1 shl h) <= max_depth) {
            h++
        }
        pr = Array<IntArray>(n, {IntArray(h){0}})
        for(i in 0 until n) {
            pr[i][0] = pv[i]
        }
        for (j in 1 until h) {
            for(i in 0 until n) {
                pr[i][j] = if(pr[i][j - 1] == -1) -1 else pr[pr[i][j - 1]][j - 1]
            }
        }
    }

    fun anc(x:Int, y:Int):Boolean {
        return (pos[x] <= pos[y] && end[y] <= end[x])
    }

    fun go_up(x_:Int, up_:Int):Int {
        var x = x_
        var up = up_
        require(pr.isNotEmpty()){"Build LCA First"}
        up = Math.min(up, (1 shl h) - 1)
        for (j in h - 1 downTo 0) {
            if (up.and(1 shl j) != 0) {
                x = pr[x][j]
                if (x == -1) {
                    break
                }
            }
        }
        return x
    }

    fun lca(x_:Int, y_:Int):Int {
        var x = x_
        var y = y_
        require(pr.isNotEmpty()){"Build LCA First"}
        if (anc(x, y)) {
            return x
        }
        if (anc(y, x)) {
            return y
        }
        if (depth[x] > depth[y]) {
            x = y.also{y = x}
        }
        for (j in h - 1 downTo 0) {
            if (pr[x][j] != -1 && !anc(pr[x][j], y)) {
                x = pr[x][j]
            }
        }
        return pr[x][0]
    }
}